clients/apple/Shared/Stateful Logic/BillingService.swift

```diff
@@ -0,0 +1,197 @@
+                                case .success(_):
+                                    await transaction.finish()
+                                    self.purchaseResult = .success
+                                case .failure(let error):
+                                    DI.errors.handleError(error)
+                                }
+                            }
+                        } else {
+                            await transaction.finish()
+                        }
+                    case .failure(let error):
+                        DI.errors.handleError(error)
+                    }
+                } catch {
+                    print("Transaction failed verification")
+                }
+            }
+        }
+    }
+
+    
+    func purchasePremium() async throws -> PurchaseResult? {
+        let accountToken = UUID()
+        let purchaseOpt: Set<Product.PurchaseOption> = [Product.PurchaseOption.appAccountToken(accountToken)]
+        
+        do {
+            guard let monthlySubscription = maybeMonthlySubscription else {
+                throw StoreError.noProduct
+            }
+            let result = try await monthlySubscription.purchase(options: purchaseOpt)
+            SKReceiptRefreshRequest().start()
+            
+            switch result {
+            case .success(let verification):
+                while true {
+                    if let receipt = getReceipt() {
+                        makingPurchaseAttempt = true
+                        let transaction = try checkVerified(verification)
+                        
+                        let result = self.core.upgradeAccountAppStore(originalTransactionId: String(transaction.originalID), appAccountToken: accountToken.uuidString.lowercased(), encodedReceipt: receipt)
+                        
+                        switch result {
+                        case .success(_):
+                            await transaction.finish()
+                            makingPurchaseAttempt = false
+                            purchaseResult = .success
+                            return .success
+                        case .failure(let error):
+                            makingPurchaseAttempt = false
+                            DI.errors.handleError(error)
+                            return .failure
+                        }
+                    }
+                }
+            case .pending:
+                purchaseResult = .pending
+                return .pending
+            case .userCancelled:
+                return nil
+            default:
+                return .failure
+            }
+        }
+        catch {
+            print("Couldn't read receipt data with error: " + error.localizedDescription)
+        }
+
+        return .failure
+    }
+    
+    func checkVerified<T>(_ result: VerificationResult<T>) throws -> T {
+        switch result {
+        case .unverified:
+            throw StoreError.failedVerification
```

Comment:
Why does this need to throw?

Reply:
Doesn't have to but if I did return `nil` I would have to add more lines of code.